File: tests/cases/classes.py
from pyjs import js

global_hello = "hello"
global_world = global_hello + " world"

@js
class Pass:
    pass

@js
class Base:

    MULTIPLIER = 2
    static_value = global_world + "!"

    def __init__(self, base_number=21):
        local_number = 9
        self._number = base_number * self.MULTIPLIER
        self.things = [
            global_hello,
            local_number,
            self.static_value,
            self._number,
            "thing"
        ]
        print("Base.__init__")
        print(self.things)

    def action(self):
        print(f"Base.action: {self._number}")
        print(f"Base.action: {self.MULTIPLIER}")

    @classmethod
    def class_action(cls):
        print(f"Base.class_action: {cls.MULTIPLIER}")

    @staticmethod
    def static_action():
        print(f"Base.static_action: {Base.MULTIPLIER}")

    def add_thing(self, thing: str):
        self.things.append(thing)
        return self

#    @property
#    def number(self):
#        return self._number
#
#    @number.setter
#    def number(self, number):
#        self._number = number

@js
def make_base():
    return Base()


@js
class SubClass(Base):
    FORTY = "forty"

    def __init__(self, two: str):
        super().__init__()
        self.word = self.FORTY + two
        self.base = make_base()
        print("SubClass.__init__")

    def do_things(self):
        if self.base.things:
            if len(self.base.things) > 1:
                print("more than 1")
            print(self.base.things[0])
        else:
            print("nothing")

    def action(self):
        super().action()
        print(f"SubClass.action: {self.word}")
        print(f"SubClass.action: {self._number}")
        print(f"SubClass.action: {self.FORTY}")
        print(f"SubClass.action: {self.MULTIPLIER}")

    @classmethod
    def class_action(cls):
        super().class_action()
        cls.static_action()
        print(f"SubClass.class_action: {cls.FORTY}")
        print(f"SubClass.class_action: {cls.MULTIPLIER}")

    @staticmethod
    def static_action():
        Base.static_action()
        print(f"SubClass.static_action: {SubClass.FORTY}")
        print(f"SubClass.static_action: {SubClass.MULTIPLIER}")


def main():
    print("Base:")
    b = Base()
    b.action()
    b.class_action()
    b.static_action()

    print("SubClass:")
    sc = SubClass("two")
    sc.do_things()
    sc.action()
    sc.class_action()
    sc.static_action()
    sc.base.static_action()

File: tests/cases/classes.js
// tests.cases.classes
const global_hello = 'hello';
class Base {
  static MULTIPLIER = 2;
  static static_value = 'hello world!';
  constructor(base_number) {
    var local_number = 9;
    this._number = (base_number * Base.MULTIPLIER);
    this.things = [global_hello, local_number, Base.static_value, this._number];
    console.log('Base.__init__');
    console.log(this.things);
  }

  action() {
    console.log('Base.action: '+this._number);
    console.log('Base.action: '+Base.MULTIPLIER);
  }

  static class_action() {
    console.log('Base.class_action: '+Base.MULTIPLIER);
  }

  static static_action() {
    console.log('Base.static_action: '+Base.MULTIPLIER);
  }

}


function make_base() {
  return new Base();
}

class SubClass extends Base {
  static FORTY = 'forty';
  constructor(two) {
    super();
    this.word = (SubClass.FORTY + two);
    this.base = make_base();
    console.log('SubClass.__init__');
  }

  do_things() {
    if (this.base.things.length) {
      if ((this.base.things.length > 1)) {
        console.log('more than 1');
      }
      console.log(this.base.things[0]);
    } else {
      console.log('nothing');
    }
  }

  action() {
    super.action();
    console.log('SubClass.action: '+this.word);
    console.log('SubClass.action: '+this._number);
    console.log('SubClass.action: '+SubClass.FORTY);
    console.log('SubClass.action: '+Base.MULTIPLIER);
  }

  static class_action() {
    super.class_action();
    SubClass.static_action();
    console.log('SubClass.class_action: '+SubClass.FORTY);
    console.log('SubClass.class_action: '+Base.MULTIPLIER);
  }

  static static_action() {
    Base.static_action();
    console.log('SubClass.static_action: '+SubClass.FORTY);
    console.log('SubClass.static_action: '+Base.MULTIPLIER);
  }

}


function main() {
  console.log('Base:');
  var b = new Base();
  b.action();
  Base.class_action();
  Base.static_action();
  console.log('SubClass:');
  var sc = new SubClass('two');
  sc.do_things();
  sc.action();
  SubClass.class_action();
  SubClass.static_action();
  Base.static_action();
}

